dipy/denoise/fast_local_covariance.pyx

```diff
@@ -0,0 +1,200 @@
+import numpy as np
+cimport cython
+cimport numpy as cnp
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef void update_outer_prod(double[:,:] out, double[:] x, int utype)nogil:
+    cdef:
+        cnp.npy_intp i, j, n
+    n = x.shape[0]
+    if utype==0:
+        for i in range(n):
+            for j in range(n):
+                out[i,j] = x[i] * x[j]
+    elif utype==1:
+        for i in range(n):
+            for j in range(n):
+                out[i,j] += x[i] * x[j]
+    elif utype==-1:
+        for i in range(n):
+            for j in range(n):
+                out[i,j] -= x[i] * x[j]
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef void update_vector(double[:] out, double[:] x, int utype)nogil:
+    cdef:
+        cnp.npy_intp i, n
+    n = x.shape[0]
+    if utype==0:
+        for i in range(n):
+            out[i] = x[i]
+    elif utype==1:
+        for i in range(n):
+            out[i] += x[i]
+    elif utype==-1:
+        for i in range(n):
+            out[i] -= x[i]
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef void update_matrix(double[:,:] out, double[:,:] cov, int utype)nogil:
+    cdef:
+        cnp.npy_intp i, j, n
+    n = cov.shape[0]
+    if utype==0:
+        for i in range(n):
+            for j in range(n):
+                out[i,j] = cov[i, j]
+    elif utype==1:
+        for i in range(n):
+            for j in range(n):
+                out[i,j] += cov[i, j]
+    elif utype==-1:
+        for i in range(n):
+            for j in range(n):
+                out[i,j] -= cov[i, j]
+
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+def fast_local_covariance(double[:,:,:,:] I, int radius, double[:,:,:,:,:] out, double[:,:,:,:] meanout):
```

Comment:
oh so at the end of the where the cython ends I will have to continue with the rest of the code, is it ?